nat-lab/tests/test_batching.py

```diff
@@ -0,0 +1,304 @@
+            ),
+            (
+                ConnectionTag.DOCKER_CONE_CLIENT_2,
+                ["stun", "local", "upnp"],
+                BatchTestFeatures(
+                    direct_interval=4,
+                    direct_threshold=0,
+                    is_meshnet=True,
+                    ip_stack=IPStack.IPv4,
+                ),
+                NO_BATCH_DOCKER_CONE_CLIENT_2_histogram_check,
+            ),
+            (
+                ConnectionTag.DOCKER_SHARED_CLIENT_1,
+                ["stun", "local", "upnp"],
+                BatchTestFeatures(
+                    direct_interval=4,
+                    direct_threshold=0,
+                    is_meshnet=False,
+                    ip_stack=IPStack.IPv4,
+                ),
+                NO_BATCH_DOCKER_SHARED_CLIENT_1_histogram_check,
+            ),
+        ])
+    ),
+]
+
+
+def BATCHED_DOCKER_CONE_CLIENT_1_histogram_check(hs: HistogramResult):
+    assert hs.outgoing[0] == 6
+    assert sum(hs.outgoing[7:8]) == 4
+    assert sum(hs.outgoing[1:7]) == 0
+
+
+def BATCHED_DOCKER_CONE_CLIENT_2_histogram_check(hs: HistogramResult):
+    assert hs.outgoing[0] == 6
+    assert sum(hs.outgoing[8:9]) == 4
+    assert sum(hs.outgoing[1:7]) == 0
+
+
+def BATCHED_DOCKER_SHARED_CLIENT_1_histogram_check(hs: HistogramResult):
+    assert hs.outgoing[0] == 6
+    assert sum(hs.outgoing[8:9]) == 4
+    assert sum(hs.outgoing[1:7]) == 0
+
+
+CLIENTS_BATCHING = [
+    pytest.param(
+        _generate_setup_parameters([
+            (
+                ConnectionTag.DOCKER_CONE_CLIENT_1,
+                ["stun", "local", "upnp"],
+                BatchTestFeatures(
+                    direct_interval=4,
+                    direct_threshold=2,
+                    is_meshnet=True,
+                    ip_stack=IPStack.IPv4,
+                ),
+                BATCHED_DOCKER_CONE_CLIENT_1_histogram_check,
+            ),
+            (
+                ConnectionTag.DOCKER_CONE_CLIENT_2,
+                ["stun", "local", "upnp"],
+                BatchTestFeatures(
+                    direct_interval=4,
+                    direct_threshold=2,
+                    is_meshnet=True,
+                    ip_stack=IPStack.IPv4,
+                ),
+                BATCHED_DOCKER_CONE_CLIENT_2_histogram_check,
+            ),
+            (
+                ConnectionTag.DOCKER_SHARED_CLIENT_1,
+                ["stun", "local", "upnp"],
+                BatchTestFeatures(
+                    direct_interval=4,
+                    direct_threshold=2,
+                    is_meshnet=False,
+                    ip_stack=IPStack.IPv4,
+                ),
+                BATCHED_DOCKER_SHARED_CLIENT_1_histogram_check,
+            ),
+        ])
+    ),
+]
+
+
+async def get_batch_histogram(
+    target: TargetMachine, mesh_node: Node
+) -> HistogramResult:
+    print(f"Running batch observer for target: {target.tag}")
+
+    await asyncio.sleep(2)
```

Comment:
it's not clear why we need both the 2s and 20s sleeps here

Reply:
20seconds - how much to capture
2seconds - how much to sleep before a third meshnet peer(gamma) is registered to the meshnet :)

Reply:
Sorry, wrong explanation and this sleep I think is not needed at all. Good catch!